go/database/mpt/proof/verification_test.go

```diff
@@ -0,0 +1,672 @@
+
+	trie := NewMockverifiableTrie(ctrl)
+	trie.EXPECT().CreateWitnessProof(gomock.Any(), gomock.Any()).Return(proof, nil).AnyTimes()
+	trie.EXPECT().VisitAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	trie.EXPECT().GetValue(gomock.Any(), gomock.Any()).Return(common.Value{}, nil).AnyTimes()
+
+	visitor := accountVerifyingVisitor{trie: trie, observer: observer, logWindow: LogWindow,
+		ctx: context.Background()}
+
+	for i := 0; i < LogWindow+1; i++ {
+		visitor.Visit(accountNode, mpt.NodeInfo{})
+	}
+}
+
+// countingWhenDoneContext is a context.Context that counts the number of times Done is called, and signals done only
+// when the threshold is reached.
+type countingWhenDoneContext struct {
+	context.Context
+	count     int // count the number of executions checking done
+	threshold int // when this threshold is reached, signal done from this point onwards
+	done      chan struct{}
+}
+
+func newCountingWhenDoneContext(ctx context.Context, threshold int) *countingWhenDoneContext {
+	return &countingWhenDoneContext{
+		Context:   ctx,
+		threshold: threshold,
+		done:      make(chan struct{}),
+	}
+}
+
+func (c *countingWhenDoneContext) Done() <-chan struct{} {
+	if c.count == c.threshold { // equality to close only once
+		close(c.done)
+	}
+	c.count++
+	return c.done
+}
+
+// errorInjectingProof is a witness.Proof that can be configured to return an error or wrong values after a certain number of calls.
+type errorInjectingProof struct {
+	witness.Proof
+	returnComplete bool
+	returnError    error
+	count          int
+	transientCount *int
+	threshold      int
+}
+
+func (p *errorInjectingProof) GetCodeHash(root common.Hash, address common.Address) (common.Hash, bool, error) {
+	if p.count >= p.threshold {
+		return common.Hash{}, p.returnComplete, p.returnError
+	}
+	p.count++
+	*p.transientCount++
+	return p.Proof.GetCodeHash(root, address)
+}
+
+func (p *errorInjectingProof) GetState(root common.Hash, address common.Address, key common.Key) (common.Value, bool, error) {
+	if p.count >= p.threshold {
+		return common.Value{}, p.returnComplete, p.returnError
+	}
+	p.count++
+	*p.transientCount++
+	return p.Proof.GetState(root, address, key)
+}
+
+func (p *errorInjectingProof) GetBalance(root common.Hash, address common.Address) (amount.Amount, bool, error) {
+	if p.count >= p.threshold {
+		return amount.Amount{}, p.returnComplete, p.returnError
+	}
+	p.count++
+	*p.transientCount++
+	return p.Proof.GetBalance(root, address)
+}
+
+func (p *errorInjectingProof) GetNonce(root common.Hash, address common.Address) (common.Nonce, bool, error) {
+	if p.count >= p.threshold {
+		return common.Nonce{}, p.returnComplete, p.returnError
+	}
+	p.count++
+	*p.transientCount++
+	return p.Proof.GetNonce(root, address)
+}
+
+// errorInjectingArchiveVerifiableTrie is a verifiableArchiveTrie that can be configured to return an error after a certain number of calls.
+type errorInjectingArchiveVerifiableTrie struct {
```

Comment:
Also for this type, what is missing in the mock library that makes a hand-crafted mock necessary?

Reply:
yes, switched to use mocks. Not sure the result is very much better, but I think it is ok